platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/event/tipset/TipsetEventCreatorImpl.java

```diff
@@ -0,0 +1,392 @@
+
+        if (beNiceToNerdChance > 0 && random.nextDouble() < beNiceToNerdChance) {
+            return createEventToReduceBullyScore();
+        } else {
+            return createEventByOptimizingTipsetScore();
+        }
+    }
+
+    /**
+     * Create an event using the other parent with the best tipset score.
+     *
+     * @return the new event, or null if it is not legal to create a new event
+     */
+    @Nullable
+    private GossipEvent createEventByOptimizingTipsetScore() {
+        final List<EventDescriptor> possibleOtherParents = childlessOtherEventTracker.getChildlessEvents();
+        Collections.shuffle(possibleOtherParents, random);
+
+        EventDescriptor bestOtherParent = null;
+        TipsetAdvancementWeight bestScore = ZERO_ADVANCEMENT_WEIGHT;
+        for (final EventDescriptor otherParent : possibleOtherParents) {
+            final TipsetAdvancementWeight parentScore =
+                    tipsetScoreCalculator.getTheoreticalAdvancementScore(List.of(otherParent));
+            if (parentScore.isGreaterThan(bestScore)) {
+                bestOtherParent = otherParent;
+                bestScore = parentScore;
+            }
+        }
+
+        if (lastSelfEvent != null && bestOtherParent == null) {
+            // There exist no parents that can advance consensus, and this is not our first event.
+            // The only time it's ok to create an event with no other parent is when we are creating
+            // our first event.
+            return null;
+        }
+
+        if (bestOtherParent != null) {
+            tipsetMetrics.getTipsetParentMetric(bestOtherParent.getCreator()).cycle();
+        }
+        return buildAndProcessEvent(bestOtherParent);
+    }
+
+    /**
+     * Create an event that reduces the bully score.
+     *
+     * @return the new event, or null if it is not legal to create a new event
+     */
+    @Nullable
+    private GossipEvent createEventToReduceBullyScore() {
+        final List<EventDescriptor> possibleOtherParents = childlessOtherEventTracker.getChildlessEvents();
+        final List<EventDescriptor> nerds = new ArrayList<>(possibleOtherParents.size());
+
+        // Choose a random nerd, weighted by how much it is currently being bullied.
+
+        // First, figure out who is a nerd and sum up all bully scores.
+        int bullyScoreSum = 0;
+        final List<Integer> bullyScores = new ArrayList<>(possibleOtherParents.size());
+        for (final EventDescriptor nerd : possibleOtherParents) {
+            final int bullyScore = tipsetScoreCalculator.getBullyScoreForNode(nerd.getCreator());
+
+            final TipsetAdvancementWeight tipsetScore =
+                    tipsetScoreCalculator.getTheoreticalAdvancementScore(List.of(nerd));
+
+            if (bullyScore > 1 && tipsetScore.isNonZero()) {
+                // Note: if bully score is greater than 1, it is mathematically not possible
```

Comment:
For things that should not be possible, maybe there should be a log statement saying that such an event occurred. Because if it does, we have a bug that needs fixing.

Reply:
Added a log statement here.

```
logger.error(EXCEPTION.getMarker(), "failed to locate eligible nerd to use as a parent");
```

Reply:
The main danger here is if this ever gets spammed thousands of times per second. IMO we should probably prioritize some sort of utility for use in logs that have the potential of being spammy.

Reply:
This isn't quite what I need, this only allows every Nth message to go through, as opposed to allowing a message to go through every N seconds.

Since this is not enabled yet, I think I will work on this as a stand alone PR.